asllib/ASLTypingReference.mld

```diff
@@ -16,107 +16,753 @@ welcome; those can be sent to Arm’s Architecture Formal Team Lead Jade
+
+  - ty is undefined.
+
+  {3 Comments}
+  This encompasses LRM Section 7.6.1. R_YZHM.
+
+{2 TypingRule.CheckUnop}
+
+  {3 Goal}
+  Checking compatibility of an unary operator with the type of its argument.
+
+  {3 Definition}
+  [t] is the result of checking compatibility of a unary operator [op] with
+  type [t1] and one of the following applies:
+  - All of the folowing applies:
+    - [op] is [BNOT];
+    - [t1] type-satisfies [boolean];
+    - [t] is [boolean];
+
+  - All of the following applies:
+    - [op] is [NEG];
+    - One of the following applies:
+      - [t1] type-satisfies [integer]; 
+      - [t1] type-satisfies [real];
+     - One of the following applies:
+       - All of the following applies:
+         - [t1] has the structure of an unconstrained integer;
+         - [t] is an unconstrained integer;
+       - All of the following applies:
+         - [t1] has the structure of a constrained integer;
+         - [t] is a constrained integer whose constraint is ;
+  
+  - All of the following applies:
+    - [op] is [NOT];
+    - [t1] has the structure of a bitvector;
+    - [t] is [t1].
+
+  {3 Comments}
+
+  {3 Examples}
+
+ {2 TypingRule.CheckBinop}
+
+  {3 Goal}
+  Checking compatibility of a binary operator with the types of its arguments.
+
+  {3 Definition}
+  [t] is the result of checking compatibility of a binary operator [op] with
+  types [t1] and [t2] and one of the following applies:
+  - All of the following applies:
+    - [op] is [AND], [OR], [EQ] or [IMPL]; 
+    - [t1] type-satisfies [boolean];
+    - [t2] type-satisfies [boolean];
+    - [t] is [boolean].
+
+  - All of the following applies:
+    - [op] is [AND], [OR], or [EOR];
+    - [t1] has the structure of a bitvector;
+    - [t2] has the structure of a bitvector;
+    - [t1] and [t2] have the same bitvector width [w];
+    - [t] is the bitvector type of width [w].
+
+  - All of the following applies:
+    - [op] is [PLUS] or [MINUS];
+    - [t1] has the structure of a bitvector;
+    - [t2] has the structure of a bitvector;
+    - [t1] and [t2] have the same bitvector width [w];
+    - [t2] type-satisfies [integer];
+    - [t] is the bitvector type of width [w].
+
+  - All of the following applies:
+    - [op] is [EQ_OP] or [NEQ];
+    - One of the following applies:
+      - [t1] is equal to [t2];
+      - All of the following applies:
+        - [t1] type-satisfies [integer];
+        - [t2] type-satisfies [integer];
+      - All of the following applies:
+        - [t1] has the structure of a bitvector;
+        - [t2] has the structure of a bitvector;
+        - [t1] and [t2] have the same bitvector width;
+      - All of the following applies:
+        - [t1] type-satisfies [boolean];
+        - [t2] type-satisfies [boolean];
+      - All of the following applies:
+        - [t1] enumerates local declarations [li1]; 
+        - [t2] enumerates local declarations [li2];
+        - [li1] equals [li2];
```

Comment:
I think we have used a shorthand for this (which is also not ideally formulated):
```suggestion
        - [t1] and [t2] both have the structure of enumeration types with the same enumeration literals;
```